File: Java-SpringBoot/BE-02-DB/java-17-springboot-3.5/src/main/java/com/flyrank/taskservice/entity/TaskEntity.java
package com.flyrank.taskservice.entity;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * JPA Entity representing a task record persisted in the database.
 */
@Entity
@Table(name = "tasks")
public class TaskEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false)
    private boolean done = false;

    @Column(name = "created_at", nullable = false, updatable = false)
    private String createdAt;

    @Column(name = "updated_at", nullable = false)
    private String updatedAt;

    public TaskEntity() {}

    public TaskEntity(String title, boolean done) {
        this.title = title;
        this.done = done;
    }

    @PrePersist
    protected void onCreate() {
        String nowStr = LocalDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME);
        if (this.createdAt == null) {
            this.createdAt = nowStr;
        }
        this.updatedAt = nowStr;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME);
    }

    // Getters and Setters
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public boolean isDone() {
        return done;
    }

    public void setDone(boolean done) {
        this.done = done;
    }

    public String getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(String createdAt) {
        this.createdAt = createdAt;
    }

    public String getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(String updatedAt) {
        this.updatedAt = updatedAt;
    }
}

File: Java-SpringBoot/BE-02-DB/kotlin-2.3-springboot-3.5/src/main/kotlin/com/flyrank/taskservice/entity/TaskEntity.kt
package com.flyrank.taskservice.entity

import jakarta.persistence.*
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * JPA Entity in Kotlin representing a task record persisted in the database.
 */
@Entity
@Table(name = "tasks")
class TaskEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,

    @Column(nullable = false)
    var title: String = "",

    @Column(nullable = false)
    var done: Boolean = false,

    @Column(name = "created_at", nullable = false, updatable = false)
    var createdAt: String? = null,

    @Column(name = "updated_at", nullable = false)
    var updatedAt: String? = null
) {
    @PrePersist
    fun onCreate() {
        val nowStr = LocalDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME)
        if (createdAt == null) {
            createdAt = nowStr
        }
        updatedAt = nowStr
    }

    @PreUpdate
    fun onUpdate() {
        updatedAt = LocalDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME)
    }
}
